src/coap_openssl.c

```diff
@@ -492,138 +510,724 @@ int coap_dtls_context_set_psk(coap_context_t *ctx,
+            if (r != X509_R_CERT_ALREADY_IN_HASH_TABLE) {
+              coap_log(LOG_WARNING,
+                       "*** setup_pki: (D)TLS: %s at %s:%s\n",
+                       ERR_reason_error_string(e),
+                       ERR_lib_error_string(e),
+                       ERR_func_error_string(e));
+            }
+          }
+        }
+      }
+      BIO_free(in);
+      X509_free(x);
     }
-    if (setup_data->private_key && setup_data->private_key[0]) {
-      if (!(SSL_CTX_use_PrivateKey_file(context->tls.ctx, setup_data->private_key, SSL_FILETYPE_PEM))) {
-        coap_log(LOG_WARNING, "*** coap_dtls_context_set_pki: TLS: %s: Unable to configure Server Private Key\n", setup_data->private_key);
+    break;
+
+  case COAP_PKI_KEY_ASN1:
+    if (setup_data->key.asn1.public_cert && setup_data->key.asn1.public_cert_len > 0) {
+      if (!(SSL_use_certificate_ASN1(ssl, setup_data->key.asn1.public_cert,
+            setup_data->key.asn1.public_cert_len))) {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure Client Certificate\n",
+                 "ASN1");
         return 0;
       }
     }
-    else if (setup_data->asn1_private_key && setup_data->asn1_private_key_len > 0) {
-      if (!(SSL_CTX_use_PrivateKey_ASN1(setup_data->asn1_private_key_type, context->tls.ctx, setup_data->asn1_private_key, setup_data->asn1_private_key_len))) {
-        coap_log(LOG_WARNING, "*** coap_dtls_context_set_pki: TLS: %s: Unable to configure Server Private Key\n", "ASN1");
+    else {
+      coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: No Client Certificate defined\n");
+      return 0;
+    }
+    if (setup_data->key.asn1.private_key &&
+             setup_data->key.asn1.private_key_len > 0) {
+      int pkey_type = map_key_type(setup_data->key.asn1.private_key_type);
+      if (!(SSL_use_PrivateKey_ASN1(pkey_type, ssl,
+            setup_data->key.asn1.private_key, setup_data->key.asn1.private_key_len))) {
+        coap_log(LOG_WARNING,
+ "*** setup_pki: (D)TLS: %s: Unable to configure Client Private Key\n",
+                 "ASN1");
         return 0;
       }
     }
     else {
-      coap_log(LOG_ERR, "*** coap_dtls_context_set_pki: TLS: No Server Private Key defined\n");
+      coap_log(LOG_ERR,
+             "*** setup_pki: (D)TLS: No Client Private Key defined\n");
+      return 0;
+    }
+    if (setup_data->key.asn1.ca_cert && setup_data->key.asn1.ca_cert_len > 0) {
+      /* Need to use a temp variable as it gets incremented*/
+      const uint8_t *p = setup_data->key.asn1.ca_cert;
+      X509* x509 = d2i_X509(NULL, &p, setup_data->key.asn1.ca_cert_len);
+      X509_STORE *st;
+      long e;
+      SSL_CTX *ctx = SSL_get_SSL_CTX(ssl);
+
+      /* We do not add CA for client to send to client as we are client! */
+
+      /* Add CA to the trusted root CA store */
+      st = SSL_CTX_get_cert_store(ctx);
+      /* Flush out existing errors */
+      while ((e = ERR_get_error()) != 0) { }
```

Comment:
For better readability, please move the closing parenthesis to the next line.

Reply:
Fixed